Rewrite custom templates article (#12775)
dotnet/docs: docs/core/tools/custom-templates.md

author: mairaw
ms.date: 08/11/2017
---

# Custom templates for dotnet new

> The [.NET Core SDK](https://www.microsoft.com/net/download/core) comes with many templates already installed and ready for you to use. The [`dotnet new` command](dotnet-new.md) isn'tonly the way to use a template, but also how to install and uninstall templates. Starting with .NET Core 2.0, you can create your own custom templates for any type of project, such as an app, service, tool, or class library. You can even create a template that outputs one or more independent files, such as a configuration file.

Reviewer: ```suggestion
The [.NET Core SDK](https://www.microsoft.com/net/download/core) comes with many templates already installed and ready for you to use. The [`dotnet new` command](dotnet-new.md) isn't only the way to use a template, but also how to install and uninstall templates. Starting with .NET Core 2.0, you can create your own custom templates for any type of project, such as an app, service, tool, or class library. You can even create a template that outputs one or more independent files, such as a configuration file.
```

```console
dotnet new --list
```

## Configuration

A template is composed of the following parts:

- Source files and folders.
- A configuration file (*template.json*).

### Source files and folders

The source files and folders include whatever files and folders you want the template engine to use when the `dotnet new <TEMPLATE>` command is run. The template engine is designed to use *runnable projects* as source code to produce projects. This has several benefits:

- The template engine doesn't require you to inject special tokens into your project's source code.
- The code files aren't special files or modified in any way to work with the template engine. So, the tools you normally use when working with projects also work with template content.
- You build, run, and debug your template projects just like you do for any of your other projects.
- You can quickly create a template from an existing project just by adding a *./.template.config/template.json* configuration file to the project.

> Files and folders stored in the template aren't limited to formal .NET project types. Source files and folders may consist of any content that you wish to create when the template is used, even if the template engine produces just one file for its output.

Reviewer: nit:

```suggestion
Files and folders stored in the template aren't limited to formal .NET project types. Source files and folders may consist of any content that you wish to create when the template is used, even if the template engine produces just one file as its output.
```

| ----------------- | ------------- | ----------- |
| `$schema`         | URI           | The JSON schema for the *template.json* file. Editors that support JSON schemas enable JSON-editing features when the schema is specified. For example, [Visual Studio Code](https://code.visualstudio.com/) requires this member to enable IntelliSense. Use a value of `http://json.schemastore.org/template`. |
| `author`          | string        | The author of the template. |
| `classifications` | array(string) | Zero or more characteristics of the template that a user might use to find the template when searching for it. The classifications also appear in the *Tags* column when it appears in a list of templates produced by using the `dotnet new -l|--list` command. |
| `identity`        | string        | A unique name for this template. |
| `name`            | string        | The name for the template that users should see. |

> | `shortName`       | string        | A default shorthand name for selecting the template that applies to environments where the template name is specified by the user, not selected via a GUI. For example, the short name is useful when using templates from a command prompt with CLI commands. |

Reviewer: Do we have naming conventions we recommend for these? If so, we should add them.

Author: Not seeing any. The schema just accepts string/array. So apparently multiples are allowed. I'll play with that 😁

```text
project_folder
│   MyDotnetTemplates.csproj
│
└───templates
    ├───mytemplate1
    │   │   console.cs
    │   │   readme.txt
    │   │
    │   └───.template.config
    │           template.json
    │
    └───mytemplate2
        │   otherfile.cs
        │
        └───.template.config
                template.json
```

## Installing a template

Use the [dotnet new -i|--install](dotnet-new.md) command to install a package.

### To install a template from a NuGet package stored at nuget.org

Use the NuGet package identifier to install a template package.

```console
dotnet new -i <NUGET_PACKAGE_ID>
```

### To install a template from a local nupkg file

Provide the path to a *.nupkg* NuGet package file.

```console
dotnet new -i <PATH_TO_NUPKG_FILE>
```

### To install a template from a file system directory

> Templates can be installed by from a template folder, such as the *mytemplate1* folder from the example above. Specify the folder path of the *.template.config* folder. The path to the template directory does not need to be absolute. However, an absolute path is required to uninstall a template that is installed from a folder.

Reviewer: ```suggestion
Templates can be installed from a template folder, such as the *mytemplate1* folder from the example above. Specify the folder path of the *.template.config* folder. The path to the template directory does not need to be absolute. However, an absolute path is required to uninstall a template that is installed from a folder.
```
